### PR TITLE
SONARJAVA-3310 Provide source maps for generated files

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

**Comment by benzonico** [comment]
LGTM !

File: java-frontend/src/main/java/org/sonar/plugins/java/api/SourceMap.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+@Beta
+public interface SourceMap {
+
+  /**
+   * Return location in JSP file corresponding to the AST node
+   *
+   * @return location in JSP file or null
+   */
+  @Nullable
+  Location sourceMapLocationFor(Tree tree);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/plugins/java/api/SourceMap.java]: What about using Optional here as well?

File: java-frontend/src/main/java/org/sonar/java/model/GeneratedFile.java [hunk 2]
```diff
@@ -27,24 +28,131 @@
...
+  public SourceMap sourceMap() {
+    if (sourceMap == null) {
+      sourceMap = new SourceMapImpl();
+    }
+    return sourceMap;
+  }
+
+  public void addSmap(SmapFile smap) {
+    if (!smap.getGeneratedFile().equals(path)) {
+      throw new IllegalStateException("Invalid smap " + smap + " for this generated file " + path);
```
[the remark this answers is not in the document]
Comment (saberduck) [reply java-frontend/src/main/java/org/sonar/java/model/GeneratedFile.java]: good point, I changed that it only prints warning into the log. I had to refactor unit test to use log tester

File: java-frontend/src/main/java/org/sonar/java/model/GeneratedFile.java [hunk 3]
```diff
@@ -27,24 +28,131 @@
...
+    }
+    smapFiles.add(smap);
+  }
+
+  class SourceMapImpl implements SourceMap {
+
+    final Map<Integer, Location> lines = new HashMap<>();
+
+    private SourceMapImpl() {
+      smapFiles.forEach(sm -> sm.getLineSection().forEach(lineInfo -> {
```
[the remark this answers is not in the document]
Comment (saberduck) [reply java-frontend/src/main/java/org/sonar/java/model/GeneratedFile.java]: I removed the lambdas, but I don't think there is a need for dedicated method in this case

**Pull request merged**